common/buildcraft/core/marker/volume/Lock.java

```diff
@@ -0,0 +1,289 @@
+        } catch (InstantiationException | IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+        cause.fromBytes(buf);
+        targets.clear();
+        IntStream.range(0, buf.readInt()).mapToObj(i -> {
+            Target target;
+            try {
+                target = new PacketBufferBC(buf).readEnumValue(Target.EnumTarget.class).clazz.newInstance();
+            } catch (InstantiationException | IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+            target.fromBytes(buf);
+            return target;
+        }).forEach(targets::add);
+    }
+
+    public static abstract class Cause {
+        public abstract NBTTagCompound writeToNBT(NBTTagCompound nbt);
+
+        public abstract void readFromNBT(NBTTagCompound nbt);
+
+        public abstract void toBytes(PacketBuffer buf);
+
+        public abstract void fromBytes(PacketBuffer buf);
+
+        public abstract boolean stillWorks(World world);
+
+        public static class CauseBlock extends Cause {
+            public BlockPos pos;
+            public Block block;
+
+            public CauseBlock() {
+            }
+
+            public CauseBlock(BlockPos pos, Block block) {
+                this.pos = pos;
+                this.block = block;
+            }
+
+            @Override
+            public NBTTagCompound writeToNBT(NBTTagCompound nbt) {
+                nbt.setTag("pos", NBTUtil.createPosTag(pos));
+                nbt.setString("block", Block.REGISTRY.getNameForObject(block).toString());
+                return nbt;
+            }
+
+            @Override
+            public void readFromNBT(NBTTagCompound nbt) {
+                pos = NBTUtil.getPosFromTag(nbt.getCompoundTag("pos"));
+                block = Block.REGISTRY.getObject(new ResourceLocation(nbt.getString("block")));
+            }
+
+            @Override
+            public void toBytes(PacketBuffer buf) {
+                buf.writeBlockPos(pos);
+                buf.writeString(Block.REGISTRY.getNameForObject(block).toString());
+            }
+
+            @Override
+            public void fromBytes(PacketBuffer buf) {
+                pos = buf.readBlockPos();
+                block = Block.REGISTRY.getObject(new ResourceLocation(buf.readString(1024)));
+            }
+
+            @Override
+            public boolean stillWorks(World world) {
+                return world.getBlockState(pos).getBlock() == block;
+            }
+        }
+
+        enum EnumCause {
+            BLOCK(CauseBlock.class);
+
+            public final Class<? extends Cause> clazz;
+
+            EnumCause(Class<? extends Cause> clazz) {
+                this.clazz = clazz;
+            }
+
+            public static Enum<EnumCause> getForClass(Class<? extends Cause> clazz) {
+                return Arrays.stream(values()).filter(enumCause -> enumCause.clazz == clazz).findFirst().orElse(null);
```

Comment:
I don't like idea of abstract method because currently registry is just this enum, and all it's values not used directly anywhere. Abstract method will just increase code size, and will be harder to add new causes/targets.